### DISCUSSION TITLE
Wayland and fractional scaling fullscreen issues

### DISCUSSION THREAD

**Discussion by hotsno** [ppy/osu#23279] [Q&A]
### Type

Cosmetic/Game behaviour

### Bug description

On Wayland (Hyprland) with fractional scaling, fullscreen mode only renders the top left 1/4 of the frame. Until I press any key, then the entire frame is rendered for some reason. Windowed and borderless work fine. Setting scaling to 1.0 (no fractional scaling) fixes the issue.

Could be a wlroots/Wayland/Hyprland issue. I also saw several similar issues but none with specifically this issue.

### Screenshots or videos

https://user-images.githubusercontent.com/71658949/233758385-b0353482-63b1-4dcd-ba9b-29584455e4e8.mp4

### Version

2023.419.0

### Logs

[database.log](https://github.com/ppy/osu/files/11300118/database.log)
[input.log](https://github.com/ppy/osu/files/11300119/input.log)
[network.log](https://github.com/ppy/osu/files/11300120/network.log)
[performance.log](https://github.com/ppy/osu/files/11300121/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11300122/runtime.log)

**Comment by Susko3** [comment]
The logs only show a bunch of
```cs
2023-04-22 03:02:08 [verbose]: SDL error log [debug]: That operation is not supported
2023-04-22 03:02:08 [verbose]: SDL error log [debug]: That operation is not supported
2023-04-22 03:02:13 [verbose]: SDL error log [debug]: That operation is not supported
```

and it was not able to update to the fullscreen resolution, only to desktop/borderless:
```cs
2023-04-22 03:01:48 [verbose]: Updated display mode to desktop resolution: 2560x1440@60, SDL_PIXELFORMAT_RGB888
```

Could you provide logs of the game working properly with 1x scaling?
Is the size of the rendered portion relative to the amount of scaling you set (1.25x, 1.5x, etc.), or is it always 1/4?

**Reply by hotsno** [reply]
Hmm I didn't get an email notification for your comment despite being "subscribed" to this discussion. Also, just out of curiosity, do you know why this was converted to a discussion while some other bug reports are kept as issues? Regardless, thanks for looking into it a bit :)

> Could you provide logs of the game working properly with 1x scaling?

[input.log](https://github.com/ppy/osu/files/11315186/input.log)
[network.log](https://github.com/ppy/osu/files/11315187/network.log)
[performance.log](https://github.com/ppy/osu/files/11315188/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11315189/runtime.log)

> Is the size of the rendered portion relative to the amount of scaling you set (1.25x, 1.5x, etc.), or is it always 1/4?

It's always 1/4.

But after more testing, I tried changing the scaling *while the game is open*, and though the game looks messed up at first, if I press Alt-Enter a few times, when it cycles back to fullscreen, it seems works as expected! Kinda. 

From my very short testing just in the menus, the resolution is a lot lower and the mouse cursor has a weird box around it. Borderless looks a lot better. This is all probably just some Wayland weirdness, not sure how useful it is to look into considering that it might be a Wayland issue, plus borderless seems to be working well for me. I mainly reported this because the game says borderless causes more input latency, so it's a shame to not have fullscreen working, though I don't even know how much this applies for Linux/Wayland.

Also, if I change back to the scaling that I had when I *opened* the game, the issue comes back. Any other scaling works.

Here are some screenshots. Note that they're in 4k so you might notice much of a difference on a lower res and/or smaller screen, but on a 32" 4K screen it is certainly noticeable :)

Fullscreen (1.5x scaling): 
![20230424_16h02m38s_grim](https://user-images.githubusercontent.com/71658949/234135128-4e81a4b2-8933-432e-b439-27bee8ea4bec.png)

Borderless:
![20230424_16h03m14s_grim](https://user-images.githubusercontent.com/71658949/234135131-d99cf33c-4b77-4030-bbb1-36ad61e613ae.png)

Fullscreen (1.0x/no scaling):
![20230424_16h13m33s_grim](https://user-images.githubusercontent.com/71658949/234135397-5d017f4a-da35-4ea9-8174-0146b1a970e5.png)

**Reply by Susko3** [reply]
The 1x scaling logs show
```
2023-04-24 22:54:17 [verbose]: Updated display mode to desktop resolution: 3840x2160@60, SDL_PIXELFORMAT_RGB888
```

1.5x logs show 2560x1440 resolution, which is exactly `3840x2160 / 1.5`. This is probably why fullscreen looks blurry, the game uses "virtual" scaled pixels instead of physical monitor pixels for rendering. On Windows, we manually set the window to use real pixels, as SDL2 isn't doing that for us. SDL3 has much better high-DPI support, so I guess it's best for us to upgrade to SDL3 at some point instead of fixing locally.

You may be able to manually disable this scaling / force it to be 1x just for osu!. This exists on Windows in Program properties → Compatibility.

**Reply by senketsu03** [reply]
The exact same thing happens for some other software when using wayland (e.g. flameshot), so I guess it's the bug on wayland side
